### DISCUSSION TITLE
[Research] Web admin with terminal access

### DISCUSSION THREAD

**Discussion by openoms** [raspiblitz/raspiblitz#3922] [Research]
https://cockpit-project.org/

Very active repo with recent developments: https://github.com/cockpit-project/cockpit

Readily available on Debian (including the RaspiOS):
`sudo apt install cockpit`

Will be available on the port 9090 over HTTPS with a self-sign certificate.
Authenticated with the linux username and password eg:
`admin`
`passswordA`

To access over the LAN allow though the firewall:
`sudo ufw allow 9090`

![image](https://user-images.githubusercontent.com/43343391/144920435-38ea187b-9406-47cf-9a3b-7367d955a52a.png)
![image](https://user-images.githubusercontent.com/43343391/144920803-550c2f50-e6de-4357-8f77-a37f3ea0aba2.png)

For remote Tor access:
`config.scripts/internet.hiddenservice.sh cockpit 443 9090`

Somewhat limited in the Tor browser due to the JavaScript blocking, but the terminal login works.

![image](https://user-images.githubusercontent.com/43343391/144919733-bf369e98-fc7c-4c92-a70d-3b86a464397e.png)

On a non-RPi Linux based system it can even be used to manage virtual machines (just like virt-manager): https://github.com/cockpit-project/cockpit-machines

And has a graphical ZFS interface: https://github.com/optimans/cockpit-zfs-manager (which seem to be EoL being picked up by the 45Drives corp https://www.45drives.com/  https://github.com/45Drives/cockpit-zfs-manager)

**Comment by openoms** [comment]
@cstenglein @rootzoll what do you think about having a terminal tab integrated to the webUI?

Same security practices should apply:
* HTTPS only (even when served over Tor)
* compare the fingerprint of the self-signed certificate on the LCD or from the menu (as already done.for webUIs)
* terminal login simply protected by the ssh password (passwordA) - as in Cockpit

Cockpit does have plugins - https://cockpit-project.org/blog/cockpit-starter-kit.html, but probably better to take the terminal part and integrate to the RaspiBlitz webUI.

A terminal in the browser window can have all the functionality not implemented in the web-interface yet and would be very handy for debugging.

Discussion links: 
https://twitter.com/openoms/status/1468902972306436099
https://bitcoinhackers.org/web/@openoms/107416707909229274

**Comment by openoms** [comment]
Cockpit TLS cert info (from the manpage):
```
To specify the TLS certificate the web service should use, simply drop
a file with the extension .cert in the /etc/cockpit/ws-certs.d
directory. If there are multiple files in this directory, then the
highest priority one is chosen after sorting.

To check which certificate cockpit-ws will use, run the following
command
```
```
$ sudo /usr/lib/cockpit/cockpit-certificate-ensure --check
/etc/cockpit/ws-certs.d/0-self-signed.cert
```

Display the SHA-1 fingerprint:
```
openssl x509 -sha1 -in /etc/cockpit/ws-certs.d/0-self-signed.cert -fingerprint -noout | cut -d"=" -f2
```
Display the SHA-256 fingerprint:
```
openssl x509 -sha256 -in /etc/cockpit/ws-certs.d/0-self-signed.cert -fingerprint -noout | cut -d"=" -f2
```

**Comment by cstenglein** [comment]
Hi, sorry for the late answer. Just writing my thoughts and concerns, don't wanna make it sound negative 😮

Fist of all, interesting idea. I'm currently not entirely convinced about the usecases though:

> A terminal in the browser window can have all the functionality not implemented in the web-interface yet 

I'm not sure what the benefit for the user would be. Can't the user just ssh into the raspiblitz then?

A cool idea would be to access non-implemented functionality on-the-go over the webui => terminal but I'm not sure how the cockpit handles the terminal output. I guess it's not mobile friendly and that would introduce another issue.

> and would be very handy for debugging.

I would say it depends. Currently, the webUI would be more targeted towards non-technical users and errors should be displayed over the webUI as well (don't know if that *really* works in practice if there is some config file that needs change etc.)

To get better debugging output, would implementing cstenglein/raspiblitz-web#49 be enough or should there some additional solutions in the webUI?

**Comment by openoms** [comment]
Thanks for your input @cstenglein !
Agree that having a terminal in the browser is mainly just a convenience factor.
I was thinking of the troubles with getting a terminal open on Windows and also on mobile, but as you say text and dialog menus are not suitable for small screens.

It actually does come with some security tradeoff also because it relies on the SSL encryption rather than ssh itself (let alone the possible browser extensions in play).

So agree, Cockpit might come useful as an option to manage the server underlying RaspiBlitz (like my case managing VMs in it), but for the RaspiBlitz WebGUI the aim should be to be able to expose the log / cli outputs.

**Comment by openoms** [comment]
I find myself using the virtual machine manager extension the most: https://github.com/cockpit-project/cockpit-machines
![image](https://user-images.githubusercontent.com/43343391/145707878-bf884fae-7ad8-4066-a7ea-933e3887b21c.png)

Components are directly linked like the Terminal:
https://localhost:9090/system/terminal

the virtual machine extension:
https://localhost:9090/machines

**Comment by openoms** [comment]
Found running cockpit very useful for testing.
It flags up the failed systemd services in the GUI and helps to keep an eye on the disk(s) too.

**Comment by cstenglein** [comment]
After having the WebUI out for some time now, I now think this would be a good idea.

Reasoning is that there are errors which you can't recover from a webUI perspective (e.g. peers.dat file from bitcoin corrupted etc.) where a terminal might make sense.

**Comment by rootzoll** [comment]
As far as I remember all solutions we found for a WebUI-Terminal just had clearnet communication between browser & server. So instead of having a fake terminal in the WebUI I would recommend a special section first that basically explains how to open a real terminal on a desktop or mobile phone. I thin kthe v1.10 milestone would a great spot for that.
